src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs

```diff
@@ -158,7 +161,15 @@ public virtual void OnSocketClosed()
         // Called on Libuv thread
         public void Tick()
         {
-            _frame.Tick();
+            if (_timeoutEnabled)
+            {
+                if (_secondsSinceLastConnectionEvent > ServerOptions.Limits.ConnectionTimeout.TotalSeconds)
+                {
+                    StopAsync();
+                }
+
+                _secondsSinceLastConnectionEvent++;
```

Comment:
It hadn't occurred to me to use a `Stopwatch` to track the time between ticks - I'll try that out. When I first wrote this I did try 3 different ways to track time more accurately but they all had significant perf costs.